firmware/cockroach_main/cockroach_main.ino

```diff
@@ -0,0 +1,98 @@
+/*  Daniel Fearn for CUSF 30/03/21 */
+
+/*  Project Cockroach
+
+    SD card SPI bus connected as follows:
+    * SS - pin 10
+    * MOSI - pin 11
+    * MISO - pin 12
+    * SCK - pin 13
+
+    MAX-M8Q gps connected as follows:
+    * TP - pin 5 (unused?)
+    * RX0 - pin 1 / TX
+    * TX0 - pin 0 / RX (default Serial pins)
+
+    Summary:
+    * Upon startup a new file named data_<index>.csv is created, where <index> is the smallest integer not already in use
+    * 5 Hz LED blink indicates an SD card error, 1 Hz blink indicates gps doesn't have a lock yet (lock should take ~30s)
+    * Each time the gps has a new valid packet, it is read
+    * Relevant data is extracted and written to the file in csv format:
+        * lat, long - degrees
+        * mean altitude above sea level - m
+        * vertical speed - m/s
+        * ground speed - m/s
+        * ground track - degrees
+        * gps time of week - ms
+        * time since startup - ms
+    * Opening and closing the file each iteration should protect the data in the event of an unplanned kinetic disassembly
+ 
+ */
+
+#include <SPI.h>
+#include <SD.h>
+#include "UBLOX.h"
+
+File datafile;
+UBLOX gps(Serial,9600);
+
+const String filenamestart = "data_";
+const String filenameend = ".csv";
+String filename;
+
+const int LED_pin = LED_BUILTIN;
+
+void setup(){
+    pinMode(LED_pin, OUTPUT);
+    if(!SD.begin(10)){
+        /* SD error, 5 Hz blink */
+        while(1){
+            digitalWrite(LED_pin, HIGH);
+            delay(100);
+            digitalWrite(LED_pin, LOW);
+            delay(100);
+        }
+    }
+
+    int index = 0;
+    while(SD.exists(filenamestart + String(index) + filenameend)){
+        index += 1;
+    }
```

Comment:
```suggestion
    while(SD.exists(filenamestart + String(index) + filenameend)) index++;
```

Reply:
This isn't very neat and scales in O(n), not great if we have loads of files on there. What I'd suggest that's quicker and would also preserve the order of files is to use UTC time from the ublox as the filename. Worst case, maybe use a more efficient search algorithm to find the highest file number.